### PR TITLE
added article about Spring Rest Docs (closes #8)

### LINKED ISSUE #8: Documenting your REST API with Spring Rest Docs
* Intro
  * what is it about?
  * why is it special?
  * what other tools are there? 
  * Mention Swagger as alternative ( https://www.tothepoint.company/blog/spring-rest-doc/)
* including Spring Rest Docs into your Integration Tests
* documenting hypermedia links
* documenting request  and response structure
* documenting field constraints
* documenting a field at a specific JSON path
* Conclusion
  * 100% test coverage of your REST endpoints

### PR DESCRIPTION
**Pull request by thombergs**

### CODE REVIEW

File: _posts/frameworks/2016-12-19-spring-restdocs.md [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            .andExpect(status().isOk())
+            .andDo(document("projects/create");
+}
+```
+
+Let's have a look at the details: `mvc()` is a helper method that creates a `MockMvc` 
+object that we use to submit a POST request to the URL `/projects`. The result of the
+request is passed into the `document()` method to automatically create documentation
+for the request. The `document()` method is statically imported from the class
+`MockMvcRestDocumentation` to keep the code readable.
```
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: I like that you explain where the method is statically imported from. Most examples skip this part, which makes it hard to follow.

File: _posts/frameworks/2016-12-19-spring-restdocs.md [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+public JUnitRestDocumentation restDocumentation = new JUnitRestDocumentation("build/generated-snippets");
+
+@Before
+public void setup() {
+    mvc = MockMvcBuilders.webAppContextSetup(applicationContext)
+            .apply(MockMvcRestDocumentation.documentationConfiguration(this.restDocumentation))
+            .build();
+}
+
+protected MockMvc mvc() {
```
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: what is this method used for?

File: _posts/frameworks/2016-12-19-spring-restdocs.md [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+This can be done using the [Asciidoctor Gradle Plugin](http://asciidoctor.org/docs/asciidoctor-gradle-plugin/)
+when you're using Gradle as your build tool or the [Asciidoctor Maven Plugin](http://asciidoctor.org/docs/asciidoctor-maven-plugin/)
+when you're using Maven. The following examples are based on Gradle.
+
+In your `build.gradle`, you will first have to define a dependency to the plugin:
+
+```groovy
+buildscript {
+    repositories {
+        mavenCentral()
```
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: just use `jcenter()` as `mavenCentral()` is included in `jcenter()` and we don't need `mavenLocal()`, right?

Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: maybe even `plugins.gradle.org` is enough

File: _posts/frameworks/2016-12-19-spring-restdocs.md [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+}
+```
+
+Next, we include the `asciidoctor` task to be run before the `jar` task, so that it is automatically
+run each build.
+
+```groovy
+jar {
+    dependsOn asciidoctor
+    from("${asciidoctor.outputDir}/html5") {
```
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: why should one include the docs inside the jar?

File: _posts/frameworks/2016-12-19-spring-restdocs.md [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+The `MockMvc` object is initialized with a `JUnitRestDocumentation` object, as shown in the next code snippet. 
+This way, the `MockMvc` object is instrumented to create [Asciidoctor](http://asciidoctor.org/)
+snippets into the folder `build/generated-snippets`.
+
+```java
+private MockMvc mvc;
+
+@Rule
+public JUnitRestDocumentation restDocumentation = new JUnitRestDocumentation("build/generated-snippets");
```
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: Line to long, for rendered page.

File: _posts/frameworks/2016-12-19-spring-restdocs.md [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+including them into a central documentation of some sorts.
+
+## Creating API Docs with Asciidoctor
+With the snippets at hand, we can now create our API documentation. The snippets are in 
+[Asciidoctor](http://asciidoctor.org/) format by default. Asciidoctor is a markup language 
+similiar to Markdown, but much more powerful. You can now simply create an Asciidoctor 
+document with your favorite text editor. That document will provide the stage for including the 
+snippets. An example document would look like this:
+
+```asciidoctor
```
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: currently there is no support for asciidoctor in rouge: https://github.com/jneen/rouge/tree/master/lib/rouge/lexers

File: _posts/frameworks/2016-12-19-spring-restdocs.md (hunk 4 above)
Comment (thombergs) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: removed packaging into the jar

File: _posts/frameworks/2016-12-19-spring-restdocs.md (hunk 5 above)
Comment (thombergs) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: not too long for desktop screen :)

Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: yet it is too long for half of 27" screen :D

File: _posts/frameworks/2016-12-19-spring-restdocs.md (hunk 6 above)
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: I created an issue 😄  https://github.com/jneen/rouge/issues/595

File: _posts/frameworks/2016-12-19-spring-restdocs.md [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    sourceDir "src/main/asciidoc"
+    outputDir "build/docs"
+}
+```
+
+Next, we include the `asciidoctor` task to be run before the `jar` task, so that it is automatically
+run each build.
+
+```groovy
+jar {
```
Comment (matthiasbalke) [inline _posts/frameworks/2016-12-19-spring-restdocs.md]: IMO better and shorter:
```groovy
build.dependsOn asciidoctor
```
